src/rustc.rs

```diff
@@ -77,6 +79,258 @@ pub fn hash_from_version_string(version: &str, index: usize) -> String {
+            }) => Ok(QualifiedToolchain {
+                channel,
+                date,
+                host: ImageArchitecture::from_target(host),
+                is_custom,
+                full,
+                sysroot,
+            }),
+            Ok(_) | Err(_) if std::env::var("CROSS_CUSTOM_TOOLCHAIN").is_ok() => {
+                QualifiedToolchain::custom(toolchain, &sysroot, msg_info)
+            }
+            Ok(_) => eyre::bail!("toolchain is not fully qualified"),
+            Err(e) => Err(e),
+        }
+    }
+}
+
+#[test]
+fn bisect() {
+    QualifiedToolchain::custom(
+        "bisector-nightly-2022-04-26-x86_64-unknown-linux-gnu",
+        "/tmp/cross/sysroot".as_ref(),
+        MessageInfo::create(true, false, None).unwrap(),
+    )
+    .unwrap();
+}
+
+#[derive(Debug, Clone, Deserialize, PartialEq, Eq)]
+pub struct Toolchain {
+    pub channel: String,
+    pub date: Option<String>,
+    pub host: Option<TargetTriple>,
+    pub is_custom: bool,
+    pub full: String,
+}
+
+impl std::fmt::Display for Toolchain {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str(&self.full)
+    }
+}
+
+impl std::str::FromStr for Toolchain {
+    type Err = eyre::Report;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        fn dig(s: &str) -> bool {
+            s.chars().all(|c: char| c.is_ascii_digit())
+        }
+        if let Some((channel, parts)) = s.split_once('-') {
+            if parts.starts_with(|c: char| c.is_ascii_digit()) {
+                // a date, YYYY-MM-DD
+                let mut split = parts.splitn(4, '-');
+                let ymd = [split.next(), split.next(), split.next()];
+                let ymd = match ymd {
+                    [Some(y), Some(m), Some(d)] if dig(y) && dig(m) && dig(d) => {
+                        format!("{y}-{m}-{d}")
+                    }
+                    _ => eyre::bail!("invalid toolchain `{s}`"),
+                };
+                Ok(Toolchain {
+                    channel: channel.to_string(),
+                    date: Some(ymd),
+                    host: split.next().map(|s| s.into()),
+                    is_custom: false,
+                    full: s.to_string(),
+                })
+            } else {
+                // channel-host
+                Ok(Toolchain {
+                    channel: channel.to_string(),
+                    date: None,
+                    host: Some(parts.into()),
+                    is_custom: false,
+                    full: s.to_string(),
+                })
+            }
+        } else {
+            Ok(Toolchain {
+                channel: s.to_string(),
+                date: None,
+                host: None,
+                is_custom: false,
+                full: s.to_string(),
+            })
+        }
+    }
+}
+
+#[derive(thiserror::Error, Debug, Clone)]
+#[error("`{0}` is not a valid toolchain")]
+pub struct ToolchainParseError(String);
```

Comment:
Is this ever used? `Toolchain::from_str` uses `Self::Err = eyre::Report`. I can safely delete it without issue.